### PR TITLE
Add separate pool for fetches

### LINKED ISSUE #520: Recovery after data loss doesn't work with more than 2 replicas per shard
This is a bit vague report, but we had an issue when one node was put back to production after a disk failure and [couldn't recover from peer replicas][1]. What happened was we put the node back to production, it started receiving parts, but the connection always broke before the part transfer was finished. There was a lot of network and disk i/o but nothing finished successfully, so the disk usage wasn't going up and the instance never recovered. Another strange thing is, when we stopped the faulty instance, the peer replicas continued to read and transmit parts (despite the receiver was down), so I had to restart the peer replicas as well to stop the disk i/o.

We did two things to mitigate the issue:
* Switch from xfs to ext4, the xfs nodes experienced strange performance degradation under load where overall load went up 8x and disk i/o almost stalled. It was also impossible to recover from replicas on xfs due to connection breaks, while it progressed from nodes on ext4.
* Change the affected node configuration to only two replicas per shard, so it is forced to recover from only one replica. This works even with xfs nodes, but makes recovery tedious.

I suspect the core of the issue is that receiving replica cannot moderate senders when it's not able to keep up which leads to connection breaks, there is currently no way to throttle throughput or have any sort of backpressure - the sender replicas just keep sending. I'm not sure how are you dealing with this, but this also affects read query performance while the replication is in progress. The secondary issue is that xfs is not really recommended as a filesystem, we're still investigating what causes this issue.

[1]: https://groups.google.com/forum/#!topic/clickhouse/uq9891JyV7M

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now `ReplicatedMergeTree` tree engines family uses a separate thread pool for replicated fetches. Size of the pool limited by setting `background_fetches_pool_size` which can be tuned with a server restart. The default value of the setting is 3 and it means that the maximum amount of parallel fetches is equal to 3 (and it allows to utilize 10G network). Fixes #520.

### CODE REVIEW

**Comment by alesapin** [comment]
```
2020.10.26 15:58:18.677241 [ 4017 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Debug> executeQuery: (from [::1]:52170) INSERT INTO r3 SELECT 94 % 10 = 3 ? 94 - 1 : 94
2020.10.26 15:58:18.677906 [ 4017 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Trace> ContextAccess (default): Access granted: INSERT(x) ON test_mv9dc3.r3
2020.10.26 15:58:18.682814 [ 4017 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Trace> ContextAccess (default): Access granted: SELECT(dummy) ON system.one
2020.10.26 15:58:18.683234 [ 4017 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2020.10.26 15:58:18.686423 [ 4104 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Debug> DiskLocal: Reserving 1.00 MiB on disk `default`, having unreserved 1.71 TiB.
2020.10.26 15:58:18.688369 [ 4104 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Debug> test_mv9dc3.r3 (d17247d1-3acd-493b-b245-0c0ca09c5320) (Replicated OutputStream): Wrote block with ID 'all_3475375757939625394_14499717038465598306', 1 rows
2020.10.26 15:58:18.715420 [ 4104 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Information> test_mv9dc3.r3 (d17247d1-3acd-493b-b245-0c0ca09c5320) (Replicated OutputStream): Block with ID all_3475375757939625394_14499717038465598306 already exists on other replicas as part all_105_105_0; will write it locally with that name.
2020.10.26 15:58:18.715620 [ 4104 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Trace> test_mv9dc3.r3 (d17247d1-3acd-493b-b245-0c0ca09c5320): Renaming temporary part tmp_insert_all_95_95_0 to all_105_105_0.
2020.10.26 15:58:18.718928 [ 4104 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Warning> test_mv9dc3.r3 (d17247d1-3acd-493b-b245-0c0ca09c5320): Tried to commit obsolete part all_105_105_0 covered by all_101_105_1 (state Committed)
2020.10.26 15:58:18.719698 [ 4104 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Trace> PipelineExecutor: Thread finished. Total time: 0.034125631 sec. Execution time: 0.000517279 sec. Processing time: 0.033589014 sec. Wait time: 1.9338e-05 sec.
2020.10.26 15:58:18.721751 [ 4017 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Information> executeQuery: Read 1 rows, 1.00 B in 0.044216729 sec., 22 rows/sec., 22.62 B/sec.
2020.10.26 15:58:18.721972 [ 4017 ] {278bf776-a847-428c-bf46-83f0cecb6747} <Debug> MemoryTracker: Peak memory usage (for query): 0.00 B.
```

Wow, errors come from inserts... Does it mean that fetches became even faster?

**Comment by alesapin** [comment]
need to add some tests

File: tests/integration/test_limited_replicated_fetches/test.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+def test_limited_fetches(started_cluster):
+    """
+        Test checks that that we utilize all available threads for fetches
+    """
+    node1.query("CREATE TABLE t (key UInt64, data String) ENGINE = ReplicatedMergeTree('/clickhouse/test/t', '1') ORDER BY tuple() PARTITION BY key")
+    node2.query("CREATE TABLE t (key UInt64, data String) ENGINE = ReplicatedMergeTree('/clickhouse/test/t', '2') ORDER BY tuple() PARTITION BY key")
+
+    with PartitionManager() as pm:
+        node2.query("SYSTEM STOP FETCHES t")
+        node1.query("INSERT INTO t SELECT 1, '{}' FROM numbers(5000)".format(get_random_string(104857)))
```
Comment (nikitamikhaylov) [inline tests/integration/test_limited_replicated_fetches/test.py:40]: Why don't you use `randomString` in ClickHouse?

**Review by nikitamikhaylov — APPROVED** [review]
I think you need to rebase on master one more time.

File: tests/integration/test_limited_replicated_fetches/test.py (hunk 1 above)
Comment (alesapin) [reply tests/integration/test_limited_replicated_fetches/test.py:40]: The new feature, I haven't used to it.

**Comment by alesapin** [comment]
No relevant failures.

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-8462.

**Pull request merged**